src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -171,25 +171,28 @@ class ChannelData {
                                       grpc_connectivity_state* state,
                                       grpc_closure* on_complete,
                                       grpc_closure* watcher_timer_init) {
-    auto* watcher = new ExternalConnectivityWatcher(
+    auto watcher = MakeRefCounted<ExternalConnectivityWatcher>(
```

Comment:
This seems a little messy, because code that is external to an `InternallyRefCounted<>` object is not really supposed to be able to create a `RefCountedPtr<>` to the object.  It works in this particular case, because we're allocating the object here rather than calling the protected `Ref()` method, but that really works only by accident -- it's not really the intended way to do this.

Is there a reasonable way to restructure this such that we take the additional refs inside of the `ExternalConnectivityWatcher` object instead of doing it here and in `RemoveExternalConnectivityWatcher()`?

One possible way to do this would be to go back to the way this was code was structured prior to #21361.  That PR changed the code to drop the mutex before calling `Start()` or `Cancel()`, since those methods hop into the WorkSerializer.  However, we could instead call those methods while holding the lock, have them acquire a new ref, and then schedule a closure on the ExecCtx before hopping into the WorkSerializer.

That approach is a little sub-optimal in that it may require two asynchronous hops (one for the ExecCtx and another for the WorkSerializer) instead of just one.  But this code isn't in the performance-critical path, so that's probably okay.

I'm also open to other approaches here if you can think of something better.

Reply:
Restructured to move all of the refs within ExternalConnectivityWatcher. Other than the horrible method signatures, it seems fine, and it also avoids having an ExecCtx jump